cpp/src/slice2cs/CsUtil.cpp

```diff
@@ -1864,212 +1865,91 @@ Slice::CsGenerator::toArrayAlloc(const string& decl, const string& sz)
 void
 Slice::CsGenerator::validateMetadata(const UnitPtr& u)
 {
-    MetadataVisitor visitor;
-    u->visit(&visitor);
-}
-
-bool
-Slice::CsGenerator::MetadataVisitor::visitUnitStart(const UnitPtr& unit)
-{
-    // Validate file metadata in the top-level file and all included files.
-    for (const auto& file : unit->allFiles())
-    {
-        DefinitionContextPtr dc = unit->findDefinitionContext(file);
-        assert(dc);
-
-        MetadataList newFileMetadata;
-        for (const auto& metadata : dc->getMetadata())
-        {
-            string_view directive = metadata->directive();
-            if (directive.find("cs:") == 0 && directive != "cs:attribute")
-            {
-                ostringstream msg;
-                msg << "ignoring invalid file metadata '" << *metadata << "'";
-                unit->warning(metadata->file(), metadata->line(), InvalidMetadata, msg.str());
-                continue;
-            }
-            newFileMetadata.push_back(metadata);
-        }
-        dc->setMetadata(std::move(newFileMetadata));
-    }
-    return true;
-}
-
-bool
-Slice::CsGenerator::MetadataVisitor::visitModuleStart(const ModulePtr& p)
-{
-    validate(p);
-    return true;
-}
-
-void
-Slice::CsGenerator::MetadataVisitor::visitClassDecl(const ClassDeclPtr& p)
-{
-    validate(p);
-}
-
-bool
-Slice::CsGenerator::MetadataVisitor::visitExceptionStart(const ExceptionPtr& p)
-{
-    validate(p);
-    return true;
-}
-
-bool
-Slice::CsGenerator::MetadataVisitor::visitStructStart(const StructPtr& p)
-{
-    validate(p);
-    return true;
-}
-
-void
-Slice::CsGenerator::MetadataVisitor::visitOperation(const OperationPtr& p)
-{
-    validate(p);
-    for (const auto& param : p->parameters())
-    {
-        visitParameter(param);
-    }
-}
-
-void
-Slice::CsGenerator::MetadataVisitor::visitParameter(const ParameterPtr& p)
-{
-    validate(p);
-}
-
-void
-Slice::CsGenerator::MetadataVisitor::visitDataMember(const DataMemberPtr& p)
-{
-    validate(p);
-}
-
-void
-Slice::CsGenerator::MetadataVisitor::visitSequence(const SequencePtr& p)
-{
-    validate(p);
-}
-
-void
-Slice::CsGenerator::MetadataVisitor::visitDictionary(const DictionaryPtr& p)
-{
-    validate(p);
-}
+    map<string, MetadataInfo> knownMetadata;
 
-void
-Slice::CsGenerator::MetadataVisitor::visitEnum(const EnumPtr& p)
-{
-    validate(p);
-}
+    // "cs:attribute"
+    MetadataInfo attributeInfo;
+    attributeInfo.validOn = {
+        typeid(Unit),
+        typeid(Module),
```

Comment:
.NET does not support attributes on C# namespaces:
https://learn.microsoft.com/en-us/dotnet/api/system.attributetargets?view=net-9.0

So this looks bogus.

Reply:
Agreed, this was the old behavior which I just tried to preserve.
I added this to the C# metadata issue I already opened: https://github.com/zeroc-ice/ice/issues/3246#issuecomment-2546350539